README
Warning: this version has crude support for xquery added and has not been properly tested or completed, use with caution.
== Standalone driver ==
This directory contains the a ruby interface to monetdb5 
written in pure ruby.

lib/MonetDB.rb
lib/MonetDBConnection.rb
lib/MonetDBStatements.rb
lib/MonetDBData.rb
lib/MonetDBExceptions.rb
lib/hasher.rb
lib/demo.rb: demo application how to interact with the database

ruby-monetdb-sql-0.1.gemspec: make file for rubygems

doc/: rubydoc in HTML format

== Installation ==

The standalone monetdb driver can be installed using the RubyGems Package Manager.

First build a gem file starting from the gemspec configuration:

$ gem build ruby-monetdb-sql-0.1.gemspec

Then install with the command:

$ gem install ruby-monetdb-sql-0.1.gem

== Usage ==
To use the standalone driver import the 'MonetDB' class and 'rubygems' (in case you installed it using gems).

A typical sequence of events is as follows:
Invoke query using the database handle to send the statement to the server and get back a result set object.

A result set object  has methods for fetching rows, moving around in the result set, obtaining column metadata, and releasing the result set.
Use a row fetching method such as fetch_row or an iterator such as each to access the rows of the result set.
If you want a count of the number of rows in the result set: invoke 'num_rows' method.
Invoke 'free' to release the result set. 

== Example ==

require 'MonetDB'

db = MonetDB.new
db.connect(user = "monetdb", passwd = "monetdb", lang = "sql", host="127.0.0.1", port = 50000, db_name = "demo", auth_type = "SHA1")

# set type_cast=true to enable MonetDB to Ruby type mapping
res = db.query("SELECT * from tables;", type_cast = false)

#puts res.debug_columns_type

puts "Number of rows returned: " + res.num_rows.to_s
puts "Number of fields: " + res.num_fields.to_s


# Get the columns' name
col_names = res.name_fields


# Iterate over the record set and retrieve on row at a time
puts res.fetch
while row = res.fetch do
  printf "%s \n", row
end

# Release the result set.
res.free

# Disconnect from server
db.close

See lib/demo.rb and the MonetDBDatar class documentation for more examples.



== ActiveRecord connector adapter ==
Active Record connects business objects and database tables to create a persistable domain model where logic and data are presented in one wrapping. It‘s an implementation of the object-relational mapping (ORM) pattern.

Required files:

adapter/lib/active_record/monetdb_adapter.rb

Usage example follows:
require 'active_record'

ActiveRecord::Base.logger = Logger.new(STDERR)
ActiveRecord::Base.colorize_logging = true

ActiveRecord::Base.establish_connection(
        :adapter => "monetdb",
        :host => "localhost",
        :database => "demo"
)

# Create a new table
class AddTests < ActiveRecord::Migration
  def self.up
    create_table :tests do |table|
        table.column :name, :string
        table.column :surname, :string
    end
  end

  def self.down
   drop_table :tests
  end

end

AddTests.up

# Migration: add a column name with a default value
class AddAge < ActiveRecord::Migration 
  def self.up
    add_column :tests, :age, :smallint, :default => 18
  end

  def self.down
    remove_column :tests, :age
  end

end

class Test < ActiveRecord::Base
end

# Insert an entry in the table
Test.create(:name => 'X', :surname => 'Y')

# add a column
AddAge.up

# return the first result of the query SELECT * from tables
row = Test.find(:first)
printf "SELECT * from tests LIMIT 1:\n"
printf "Name: %s, Surname: %s, Age: %s\n", row.name, row.surname, row.age

# Drop the table
AddTests.down

== Rubygem ==

The standalone ruby driver can be distributed as a ruby gem.
A gem file is already available; however, it can be generated 
starting from the ruby-monetdb-sql-0.1.gemspec file:

$ gem build ruby-monetdb-sql-0.1.gemspec

To install the file run the command:

$ gem install ruby-monetdb-sql-0.1.gem

Documentation in ri and html format will be generated and installed as well



Changelog
* Thu 19 Nov 2009 21:10 CET
Added crude support for XQuery

* Tue 30 Jun 2009 22:14:38 CEST
Added test cases for the standalone driver (ruby unit test).

* Thu 25 Jun 2009 17:31:02 CEST
Fixed a major bug that resulted in data corruption;
MAPI implementation code cleanup and bug fixes;
Support to ruby 1.9;
Both the standalone driver and activerecord now support transactions; nested transactions
are simulated via savepoints in activerecord.
Added a Rakefile and script to run activerecord unit test suite.
Type conversion in the standalone driver is now performed after data has been retrieved and can be executed on single fields.

* Mon 25 May 2009 17:52:01 CEST
Imported last week changes (protocol 9 support, parametrized the connection options in the activerecord adapter, fixed a bug in the auth protocol v8).
Fixed a bug in the mapi protocol that resulted in data loss (wrong handling of TRANSACTIONS).
Began to port the activerecord test suite to monetdb (not all test cases can be performed).
Removed an unneeded file ('lib/MonetDBStatement.rb') from the gemspec of the standalone driver (the feature will be moved to HEAD).
Began to port the driver to ruby 1.9.
Removed *.gem files from cvs.

* Mon 18 May 2009 15:22:31 CEST
Fixed bugs that prevented the correct working of activerecords' migration;
The activerecord connector now supports insertion, update and alter table operations;
Type casting is working in activerecord;
Added a rubygem and rakefile for activerecord-monetdb-adapter;
Added a new usage example for activerecord to the README file;
Added an example directory to the cvs tree;
The driver now correctly works with merovingian.


* Sat  9 May 2009 15:58:36 CEST
Fixed bugs with the query processing in the standalone driver;
Added INSERT and UPDATE methods in the activerecord connector.

* Thu  7 May 2009 17:03:01 CEST

Added a check against the protocol version during authentication;
Imported the activerecord code (available under adapter/).


lib/MonetDB.rb
# The contents of this file are subject to the MonetDB Public License
# Version 1.1 (the "License"); you may not use this file except in
# compliance with the License. You may obtain a copy of the License at
# http://monetdb.cwi.nl/Legal/MonetDBLicense-1.1.html
#
# Software distributed under the License is distributed on an "AS IS"
# basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
# License for the specific language governing rights and limitations
# under the License.
#
# The Original Code is the MonetDB Database System.
#
# The Initial Developer of the Original Code is CWI.
# Portions created by CWI are Copyright (C) 1997-July 2008 CWI.
# Copyright August 2008-2009 MonetDB B.V.
# All Rights Reserved.
#
# This file has been edited by Tinco Andringa to support XQuery


# A typical sequence of events is as follows:
# Fire a query using the database handle to send the statement to the server and get back a result set object.

# A result set object  has methods for fetching rows, moving around in the result set, obtaining column metadata, and releasing the result set.
# Use a row fetching method such as 'fetch_row' or an iterator such as each to access the rows of the result set.
# Call 'free' to release the result set. 

#module MonetDB

  require 'MonetDBConnection'
  require 'MonetDBData'
  require 'MonetDBExceptions'
  
    # = Introduction
    #
    # A typical sequence of events is as follows:
    # Create a database instance (handle), invoke query using the database handle to send the statement to the server and get back a result set object.
    #
    # A result set object  has methods for fetching rows, moving around in the result set, obtaining column metadata, and releasing the result set.
    # A result set object is an instance of the MonetDBData class.
    #
    # Records can be returneds as arrays and iterators over the set.
    #
    # A database handler (dbh) is and instance of the MonetDB class.
    #
    # = Connection management 
    #
    #  connect    -  establish a new connection
    #                * user: username (default is monetdb)
    #                * passwd: password (default is monetdb)
    #                * lang: language (default is sql) 
    #                * host: server hostanme or ip  (default is localhost)
    #                * port: server port (default is 50000)
    #                * db_name: name of the database to connect to
    #                * auth_type: hashing function to use during authentication (default is SHA1)
    #
    #  is_connected? - returns true if there is an active connection to a server, false otherwise 
    #  reconnect     - recconnect to a server
    #  close         - terminate a connection
    #  auto_commit?  - returns ture if the session is running in auto commit mode, false otherwise  
    #  auto_commit   - enable/disable auto commit mode.
    #
    #  query         - fire a query
    #
    # Currently MAPI protocols 8 and 9 are supported.
    #
    # = Managing record sets 
    #
    #
    # A record set is represented as an instance of the MonetDBData class; the class provides methods to manage retrieved data.
    #
    #
    # The following methods allow to iterate over data:
    # 
    # fetch          - iterates over the record set and retrieves on row at a time. Each row is returned as an array.
    # fetch_hash     - iterates over columns (on cell at a time). 
    # fetch_all_hash - returns record set entries hashed by column name orderd by column position.
    # 
    # To return the record set as an array (with each tuple stored as array of fields) the following method can be used:
    # 
    # fetch_all      - fetch all rows and store them  
    #
    #
    # Information about the retrieved record set can be obtained via the following methods:
    #
    # num_rows       - returns the number of rows present in the record set
    # num_fields     - returns the number of fields (columns) that compose the schema
    # name_fields    - returns the (ordered) name of the schema's columns
    # type_fields    - returns the (ordered) types list of the schema's columns
    #
    # To release a record set MonetDBData#free can be used.
    #
    # = Type conversion
    #
    # A mapping between SQL and ruby type is supported. Each retrieved field can be converted to a ruby datatype via 
    # a getTYPE method.
    #
    # The currently supported cast methods are:
    #
    # getInt        - convert to an integer value
    # getFloat      - convert to a floating point value
    # getString     - return a string representation of the value, with trailing and leading " characters removed
    # getBlob       - convert an SQL stored HEX string to its binary representation
    # getTime       - return a string representation of a TIME field
    # getDate       - return a string representation of a DATE field
    # getDateTime   - convert a TIMESTAMP field to a ruby Time object
    # getChar       - on ruby >= 1.9, convert a CHAR field to char 
    # getBool       - convert a BOOLEAN field to a ruby bool object. If the value of the field is unknown, nil is returned
    # getNull       - convert a NULL value to a nil object
    #
    #
    # = Transactions 
    #
    # By default monetdb works in auto_commit mode. To turn this feature off MonetDB#auto_commit(flag=false) can be used.
    #
    # Once auto_commit has been disable it is possible to start transactions, create/delete savepoints, rollback and commit with 
    # the usual SQL statements.
    #
    # Savepoints IDs can be generated using the MonetDB#save method. To release a savepoint ID use MonetDB#release.
    # 
    # Savepoints can be accessed (as a stack) with the MonetDB#transactions method.
    #
    # example/standalone.rb contains usage example of the above mentioned methods.

    # = Introduction
    #
    # A typical sequence of events is as follows:
    # Create a database instance (handle), invoke query using the database handle to send the statement to the server and get back a result set object.
    #
    # A result set object  has methods for fetching rows, moving around in the result set, obtaining column metadata, and releasing the result set.
    # A result set object is an instance of the MonetDBData class.
    #
    # Records can be returneds as arrays and iterators over the set.
    #
    # A database handler (dbh) is and instance of the MonetDB class.
    #
    # = Connection management 
    #
    #  connect    -  establish a new connection
    #                * user: username (default is monetdb)
    #                * passwd: password (default is monetdb)
    #                * lang: language (default is sql) 
    #                * host: server hostanme or ip  (default is localhost)
    #                * port: server port (default is 50000)
    #                * db_name: name of the database to connect to
    #                * auth_type: hashing function to use during authentication (default is SHA1)
    #
    #  is_connected? - returns true if there is an active connection to a server, false otherwise 
    #  reconnect     - recconnect to a server
    #  close         - terminate a connection
    #  auto_commit?  - returns ture if the session is running in auto commit mode, false otherwise  
    #  auto_commit   - enable/disable auto commit mode.
    #
    #  query         - fire a query
    #
    # Currently MAPI protocols 8 and 9 are supported.
    #
    # = Managing record sets 
    #
    #
    # A record set is represented as an instance of the MonetDBData class; the class provides methods to manage retrieved data.
    #
    #
    # The following methods allow to iterate over data:
    # 
    # fetch          - iterates over the record set and retrieves on row at a time. Each row is returned as an array.
    # fetch_hash     - iterates over columns (on cell at a time). 
    # fetch_all_hash - returns record set entries hashed by column name orderd by column position.
    # 
    # To return the record set as an array (with each tuple stored as array of fields) the following method can be used:
    # 
    # fetch_all      - fetch all rows and store them  
    #
    #
    # Information about the retrieved record set can be obtained via the following methods:
    #
    # num_rows       - returns the number of rows present in the record set
    # num_fields     - returns the number of fields (columns) that compose the schema
    # name_fields    - returns the (ordered) name of the schema's columns
    #
    # To release a record set MonetDBData#free can be used.
    #
    # = Type conversion 
    #
    # Invoking MonetDB#query with the flag type_conversion=true will result in a type cast of the record set fields from SQL types to ruby types
    #
    # demo.rb contains usage example of the above mentioned methods.

  class MonetDB
    def initalize()
      @connection = nil 
    end
  
    # Establish a new connection.
    #                * user: username (default is monetdb)
    #                * passwd: password (default is monetdb)
    #                * lang: language (default is sql) 
    #                * host: server hostanme or ip  (default is localhost)
    #                * port: server port (default is 50000)
    #                * db_name: name of the database to connect to
    #                * auth_type: hashing function to use during authentication (default is SHA1)
    def connect(username = "monetdb", password = "monetdb", lang = "sql", host="127.0.0.1", port = "50000", db_name = "test", auth_type = "SHA1")
      # TODO: handle pools of connections
      @username = username
      @password = password
      @lang = lang
      @host = host
      @port = port
      @db_name = db_name
      @auth_type = auth_type
      @connection = MonetDBConnection.new(user = @username, passwd = @password, lang = @lang, host = @host, port = @port)
      @connection.connect(@db_name, @auth_type)
    end
  
    # Send a <b> user submitted </b> query to the server and store the response.
    # Returns and instance of MonetDBData.
    def query(q = "")
      if  @connection != nil
        @data = MonetDBData.new(@connection, @lang)
        puts @data.execute(q)
      end
      return @data
    end
    
    # Return true if there exists a "connection" object
    def is_connected?
      if @connection == nil
        return false
      else 
        return true
      end
    end
  
    # Reconnect to the server
    def reconnect
      if @connection != nil
        self.close
        
        @connection = MonetDBConnection.new(user = @username, passwd = @password, lang = @lang, host = @host, port = @port)
        @connection.connect(db_name = @db_name, auth_type = @auth_type)
      end
    end
    
    # Turn auto commit on/off
    def auto_commit(flag=true)
      @connection.set_auto_commit(flag)
    end
  
    # Returns the current auto commit  (on/off) settings.
    def auto_commit?
      @connection.auto_commit?
    end
    
    # Returns the name of the last savepoint in a transactions pool
    def transactions
      @connection.savepoint
    end
    
    # Create a new savepoint ID
    def save
      @connection.transactions.save
    end
    
    # Release a savepoint ID
    def release
      @connection.transactions.release
    end
    
    # Close an active connection
    def close()
      @connection.disconnect
      @connection = nil
    end
  end
#end
